feat: adding BeagleView states to iOS doc (#314)
ZupIT/docs-beagle: content/en/resources/customization/beagle-for-ios/beagle-view.md

Beagle View is an `UIView` that must be used when it is necessary to add a Beagle's component in a native screen that has other `views` and use `AutoLayout`.

It is possible to initialize a `BeagleView` with a Beagle **component** or **screenType**

**Component:** Receives a declarative component. 

> **ScreenType.Remote:** Receives an URL with a server-driven component. This initialization allows you to configure BeagleView's states through `beagleViewStateObserver` parameter.

Reviewer: ```suggestion
**ScreenType.Remote:** Receives a URL with a server-driven component. This initialization allows you to configure BeagleView's states through `beagleViewStateObserver` parameter.
```

**ScreenType:** It's an enum that allows you to initialize BeagleView in three ways:

> * `Remote:` Receives an URL with a server-driven component. If there is an error when loading your server-driven screen, an optional fallback view can be configured, which would be a screen build on the frontend.

Reviewer: ```suggestion
* `Remote:` Receives a URL with a server-driven component. If there is an error when loading your server-driven screen, an optional fallback view can be configured, which would be a screen built on the frontend.
```

  You can also configure a header through `additionalData` parameter.

* `Declarative:` Receives a declarative screen (created on your frontend side on Swift).

* `Remote:` Receives a screen in a JSON format.

```swift
public class BeagleView: UIView {
    // MARK: - Initialization
    
    public convenience init(_ component: ServerDrivenComponent) {
        self.init(.declarative(component.toScreen()))
    }
    
    public convenience init(_ screenType: ScreenType) {
            self.init(viewModel: .init(screenType: screenType))
    }

    public convenience init(_ remote: ScreenType.Remote, beagleViewStateObserver: @escaping BeagleViewStateObserver) {
        self.init(viewModel: .init(screenType: .remote(remote), beagleViewStateObserver: beagleViewStateObserver))
    }
}

```

## Using Beagle View

> Check out below, a native layout that will be used to exemplify the use of `AutoLayout` together with Beagle:

Reviewer 2: ```suggestion
Check out below a native layout to exemplify the use of `AutoLayout` together with Beagle:
```

```swift
// 1   
private lazy var beagleViewDeclarative = BeagleView(
    Text("I'm a beagle text")
)

// 2 
private lazy var beagleViewRemote = BeagleView(
    .remote(.init(url: "your URL"))
)

private func setupDeclarative() {
    view.addSubview(beagleViewDeclarative)
    beagleViewDeclarative.translatesAutoresizingMaskIntoConstraints = false
    beagleViewDeclarative.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
    beagleViewDeclarative.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
    beagleViewDeclarative.topAnchor.constraint(equalTo: titleScreen.bottomAnchor, constant: 5).isActive = true
}

private func setupRemote() {
    view.addSubview(beagleViewRemote)
    beagleViewRemote.translatesAutoresizingMaskIntoConstraints = false
    beagleViewRemote.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
    beagleViewRemote.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
    beagleViewRemote.topAnchor.constraint(equalTo: beagleViewDeclarative.bottomAnchor, constant: 5).isActive = true
}

```

### 3. Error Handling

> You can observe the states and deal with errors from the `beagleViewStateObserver `parameter at BeagleView startup with the type `ScreenType.Remote`.
> 
> Check out below an example where a BeagleView was created and it receives BeagleView's states to treat them.

Reviewer 2: ```suggestion
You can see the states and deal with the `beagleViewStateObserver` parameter errors at the BeagleView startup with the `ScreenType.Remote` type.

Check out an example where a BeagleView was created and it receives BeagleView's states to treat them.
```

Author: Acredito que mudou um pouco o significado e ficou confuso, mas com base na sua sugestão, o que você acha dessa forma: 

`You can observe BeagleViews's states and deal with errors through the `beagleViewStateObserver` parameter at BeagleView startup with the `ScreenType.Remote` type.
`

Reviewer 2: -- ótimoo, agora entendi o sentido ;)

```swift
private lazy var beagleViewStateful = BeagleView(.init(url: "your URL")) { state in
    switch state {
    case .started:
        self.loadingLabel.isHidden = false
        self.errorLabel.isHidden = true
    case .error(var serverDrivenError, let retry):
        self.errorLabel.text = serverDrivenError.localizedDescription
        self.errorLabel.textColor = .red
        self.errorLabel.isHidden = false
    case .success:
        self.errorLabel.isHidden = true
    case .finished:
        self.loadingLabel.isHidden = true
    }
}

private func setupStatefulBeagleView() {
    view.addSubview(beagleViewStateful)
    beagleViewStateful.translatesAutoresizingMaskIntoConstraints = false
    beagleViewStateful.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
    beagleViewStateful.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
    beagleViewStateful.topAnchor.constraint(equalTo: beagleViewRemote.bottomAnchor, constant: 5).isActive = true
}
```

### 4. Complete Example

> You can see a complete `BeagleViewViewController` class example below:

Reviewer 2: ```suggestion
You can see a `BeagleViewViewController` class example below:
```

        beagleViewDeclarative.topAnchor.constraint(equalTo: titleScreen.bottomAnchor, constant: 5).isActive = true
    }

    private func setupRemote() {
        view.addSubview(beagleViewRemote)
        beagleViewRemote.translatesAutoresizingMaskIntoConstraints = false
        beagleViewRemote.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
        beagleViewRemote.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
        beagleViewRemote.topAnchor.constraint(equalTo: beagleViewDeclarative.bottomAnchor, constant: 5).isActive = true
    }

    private func setupStatefulBeagleView() {
        view.addSubview(beagleViewStateful)
        beagleViewStateful.translatesAutoresizingMaskIntoConstraints = false
        beagleViewStateful.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
        beagleViewStateful.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
        beagleViewStateful.topAnchor.constraint(equalTo: beagleViewRemote.bottomAnchor, constant: 5).isActive = true
    }
    
}
```

> Pronto, você já pode usar a `BeagleView`.

Reviewer: Shouldn't this be in english?

Author: yep 😅 I was gonna commit that in the next commit

Beagle View is an `UIView` that must be used when it is necessary to add a Beagle's component in a native screen that has other `views` and use `AutoLayout`.

It is possible to initialize a `BeagleView` with a Beagle **component** or **screenType**

**Component:** Receives a declarative component. 

> **ScreenType.Remote:** Receives an URL with a server-driven component. This inicialization allows you to configure BeagleView's states through `beagleViewStateObserver` parameter.

Reviewer 2: This initialization

> **ScreenType:** It's a enum that allows you to initialize BeagleView in three ways:

Reviewer 2: It's an enum

> * `Remote:` Receives an URL with a server-driven component. If there was an error when loading your server-driven screen, an optional fallback view can be configured, which would be a screen build on the frontend.

Reviewer 2: If there is an error ...

  You can also configure a header through `additionalData` parameter.

* `Declarative:` Receives a declarative screen (created on your frontend side on Swift).

* `Remote:` Receives a screen in a JSON format.

```swift
public class BeagleView: UIView {
    // MARK: - Initialization
    
    public convenience init(_ component: ServerDrivenComponent) {
        self.init(.declarative(component.toScreen()))
    }
    
    public convenience init(_ screenType: ScreenType) {
            self.init(viewModel: .init(screenType: screenType))
    }

    public convenience init(_ remote: ScreenType.Remote, beagleViewStateObserver: @escaping BeagleViewStateObserver) {
        self.init(viewModel: .init(screenType: .remote(remote), beagleViewStateObserver: beagleViewStateObserver))
    }
}

```

## Using Beagle View

> Below is a native layout that will be used to exemplify the use of `AutoLayout` together with Beagle:

Reviewer 2: Check out below, a native layout ...

### 1. Create a BeagleViewViewController

Create a class of type `UIViewController` as in the following example.

```swift
class BeagleViewViewController: UIViewController {

    // MARK: Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
    
    }
}
```

Agora, você irá criar três textos nativos do tipo `UILabel` e irá posicioná-los com o `AutoLayout`, como na função `setupLabels` abaixo.

> Now, inside `BeagleViewViewController` class, three native `UILabel` texts will be created and positioned with `AutoLayout`, as can be seen in the `setupLabels` function.

Reviewer 2: Now, inside `BeagleViewViewController` class, three native `UILabel` texts will be created and positioned with `AutoLayout`, **as you can see in the** .....

    view.addSubview(titleScreen)
    titleScreen.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
    titleScreen.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
    if #available(iOS 11.0, *) {
        titleScreen.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 5).isActive = true
    } else {
        titleScreen.topAnchor.constraint(equalTo: view.topAnchor, constant: 5).isActive = true
    }

    view.addSubview(loadingLabel)
    loadingLabel.topAnchor.constraint(equalTo: titleScreen.bottomAnchor, constant: 30).isActive = true
    loadingLabel.centerXAnchor.constraint(equalTo: titleScreen.centerXAnchor).isActive = true

    view.addSubview(errorLabel)
    errorLabel.topAnchor.constraint(equalTo: titleScreen.bottomAnchor, constant: 30).isActive = true
    errorLabel.leadingAnchor.constraint(lessThanOrEqualTo: view.leadingAnchor).isActive = true
    errorLabel.trailingAnchor.constraint(lessThanOrEqualTo: view.trailingAnchor).isActive = true
    errorLabel.centerXAnchor.constraint(equalTo: titleScreen.centerXAnchor).isActive = true
}
```

### 2. Create a BeagleView

> Below are two examples of how to instantiate a `BeagleView`, one starting with a **declarative** Beagle component and the other in **remote** mode with a server-driven component.

Reviewer 2: See below two examples on how to instantiate a `BeagleView`, one starting with a **declarative** Beagle component and the other in a **remote** mode with a server-driven component.

    .remote(.init(url: "your URL"))
)

private func setupDeclarative() {
    view.addSubview(beagleViewDeclarative)
    beagleViewDeclarative.translatesAutoresizingMaskIntoConstraints = false
    beagleViewDeclarative.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
    beagleViewDeclarative.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
    beagleViewDeclarative.topAnchor.constraint(equalTo: titleScreen.bottomAnchor, constant: 5).isActive = true
}

private func setupRemote() {
    view.addSubview(beagleViewRemote)
    beagleViewRemote.translatesAutoresizingMaskIntoConstraints = false
    beagleViewRemote.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
    beagleViewRemote.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
    beagleViewRemote.topAnchor.constraint(equalTo: beagleViewDeclarative.bottomAnchor, constant: 5).isActive = true
}

```

### 3. Error Handling

You can observe the states and deal with errors from the `beagleViewStateObserver `parameter at BeagleView startup with the type `ScreenType.Remote`.

> Below is an example where a BeagleView was created that receives BeagleView's states to be able to treat them.

Reviewer 2: Check out below an example where a BeagleView was created and it receives BeagleView's states to treat them.

```swift
private lazy var beagleViewStateful = BeagleView(.init(url: "your URL")) { state in
    switch state {
    case .started:
        self.loadingLabel.isHidden = false
        self.errorLabel.isHidden = true
    case .error(var serverDrivenError, let retry):
        self.errorLabel.text = serverDrivenError.localizedDescription
        self.errorLabel.textColor = .red
        self.errorLabel.isHidden = false
    case .success:
        self.errorLabel.isHidden = true
    case .finished:
        self.loadingLabel.isHidden = true
    }
}

private func setupStatefulBeagleView() {
    view.addSubview(beagleViewStateful)
    beagleViewStateful.translatesAutoresizingMaskIntoConstraints = false
    beagleViewStateful.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
    beagleViewStateful.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
    beagleViewStateful.topAnchor.constraint(equalTo: beagleViewRemote.bottomAnchor, constant: 5).isActive = true
}
```

### 4. Complete Example

> Bellow is the complete example of `BeagleViewViewController` class.

Reviewer 2: You can see a complete `BeagleViewViewController` class example below:
